weed/filer/postgres3/postgres3_store.go

```diff
@@ -0,0 +1,512 @@
+package postgres3
+
+/*
+ * Copyright 2022 Splunk Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"os"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/lib/pq"
+	_ "github.com/lib/pq"
+	"github.com/seaweedfs/seaweedfs/weed/filer"
+	"github.com/seaweedfs/seaweedfs/weed/filer/abstract_sql"
+	"github.com/seaweedfs/seaweedfs/weed/glog"
+	"github.com/seaweedfs/seaweedfs/weed/pb/filer_pb"
+	"github.com/seaweedfs/seaweedfs/weed/util"
+)
+
+const (
+	CONNECTION_URL_PATTERN = "host=%s port=%d sslmode=%s connect_timeout=30"
+
+	createTablePattern = `CREATE TABLE IF NOT EXISTS "%s" (
+		key       varchar(65535) PRIMARY KEY,
+		name      varchar(65535),
+		prefixes  bigint[],
+		meta      bytea
+	)`
+	createTableIndexPattern = `CREATE INDEX on "%s" USING gin (prefixes);`
+	deleteTablePattern      = `DROP TABLE "%s";`
+	insertEntryPattern      = `INSERT INTO "%s" (key, name, prefixes, meta) VALUES ($1, $2, $3, $4)
+	ON CONFLICT (key)
+	DO
+	  UPDATE SET meta = EXCLUDED.meta;`
+	findEntryPattern      = `SELECT meta FROM "%s" WHERE key = $1`
+	deleteEntryPattern    = `DELETE FROM "%s" WHERE key = $1`
+	listEntryQueryPattern = `SELECT key, name, isdir, meta FROM
+	(
+		SELECT key, name, false as isdir, meta FROM "%s"
+		WHERE prefixes @> $1 AND cardinality(prefixes) < $2 - 1
+		AND name __COMPARISON__ $3 AND name LIKE $4 ORDER BY key ASC LIMIT $6
+	) s1
+	UNION
+	(
+		SELECT dir, dir, true isdir, NULL::bytea meta FROM
+		(
+			SELECT DISTINCT split_part(key, '/', $2) AS dir FROM "%s"
+			WHERE prefixes @> $1 AND cardinality(prefixes) > $5 - 1 ORDER BY dir ASC
+		) t1
+		WHERE t1.dir > $3 AND t1.dir LIKE $4 ORDER BY dir ASC
+	)
+	ORDER BY name ASC LIMIT $6`
+	deleteFolderChildrenPattern = `DELETE FROM "%s" WHERE prefixes @> $1 and key like $2`
+)
+
+var (
+	listEntryExclusivePattern string
+	listEntryInclusivePattern string
+)
+
+var _ filer.BucketAware = (*Postgres3Store)(nil)
+
+func init() {
+	filer.Stores = append(filer.Stores, &Postgres3Store{})
+
+	listEntryExclusivePattern = strings.ReplaceAll(listEntryQueryPattern, "__COMPARISON__", ">")
+	listEntryInclusivePattern = strings.ReplaceAll(listEntryQueryPattern, "__COMPARISON__", ">=")
+}
+
+type Postgres3Store struct {
+	DB                 *sql.DB
+	SupportBucketTable bool
+	dbs                map[string]bool
+	dbsLock            sync.Mutex
+}
+
+func (store *Postgres3Store) GetName() string {
+	return "postgres3"
```

Comment:
how about "postgresForS3"?

Reply:
Sure, maybe with underscores like `postgres_for_s3` or `postgres_s3`?

Reply:
`postgres_s3` should be clear enough.

Reply:
Renamed to `postgres_s3`